### ISSUE TITLE
please get proxy settings for intersphinx

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#122]
Hello,
working behind a Squid proxy I get the following error with intersphinx:

```
#!python

WARNING: intersphinx inventory
 'http://docs.python.org/objects.inv' not fetchable due 
to <type 'exceptions.IOError'>: [Errno socket error]
 (10060, 'Operation timed out')
```

Maybe use
urllib2.getproxies()

 # see: http://thread.gmane.org/gmane.comp.python.tutor/51769/focus=51774

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/122
- Originally reported by: [timmie](https://bitbucket.org/timmie)
- Originally created at: 2009-03-10T14:43:25.086

**Comment by shimizukawa** [comment]
_From Anonymous on 2009-03-11 11:02:57+00:00_

Hello,
the code below fixed the issue for me:

{{{
#!python

import time
import urllib
import urllib2
import posixpath
from os import path

from docutils import nodes

from sphinx.builder import INVENTORY_FILENAME

def fetch_inventory(app, uri, inv):
    """Fetch, parse and return an intersphinx inventory file."""
    invdata = {}
    # both _uri_ (base URI of the links to generate) and _inv_ (actual
    # location of the inventory file) can be local or remote URIs
    localuri = uri.find('://') == -1
    #--- get proxy configuration
    proxy_settings = urllib2.getproxies()

# print proxy_settings

```
# note that in this case only FTP proxy was set.
proxy_address = proxy_settings.items()[0][1]
```

# print proxy_address

```
proxy = {'http' :  'http://'+proxy_address}  
```

# print  proxy

```
try:
    if inv.find('://') != -1:
        f = urllib.urlopen(inv, proxies=proxy)
    else:
        f = open(path.join(app.srcdir, inv))
except Exception, err:
    app.warn('intersphinx inventory %r not fetchable due to '
             '%s: %s' % (inv, err.__class__, err))
    return
try:
    line = f.next()
    if line.rstrip() != '# Sphinx inventory version 1':
        raise ValueError('unknown or unsupported inventory version')
    line = f.next()
    projname = line.rstrip()[11:].decode('utf-8')
    line = f.next()
    version = line.rstrip()[11:]
    for line in f:
        name, type, location = line.rstrip().split(None, 2)
        if localuri:
            location = path.join(uri, location)
        else:
            location = posixpath.join(uri, location)
        invdata[name] = (type, projname, version, location)
    f.close()
except Exception, err:
    app.warn('intersphinx inventory %r not readable due to '
             '%s: %s' % (inv, err.__class__, err))
else:
    return invdata
```

}}}

Please include it in the Sphinx code.

**Comment by shimizukawa** [comment]
_From Sebastian Wiesner on 2009-03-19 14:50:49+00:00_

Changeset attached.  It implements proxy support by installing a custom urllib2 opener.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-03-24 09:03:02+00:00_

Applied in <<cset 73942678c89a94db33831ec79e0441e928f4b305>>.

**Comment by shimizukawa** [comment]
_From [timmie](https://bitbucket.org/timmie) on 2009-05-05 10:11:19+00:00_

Today I got this:

WARNING: intersphinx inventory 'http://matplotlib.sourceforge.net/objects.inv' not fetchable due to <class 'urllib2.URLError'>: <urlopen error (10060, 'Operation timed out')>
WARNING: intersphinx inventory 'http://docs.python.org/dev/objects.inv' not fetchable due to <class 'urllib2.URLError'>: <urlopen error (10060, 'Operation timed out')>
WARNING: intersphinx inventory 'http://pytseries.sourceforge.net/objects.inv' not fetchable due to <class 'urllib2.URLError'>: <urlopen error (10060, 'Operation timed out')>

Using 0.61

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-05-10 19:10:26+00:00_

So, did it work for you before? What did you change so that it didn't?

**Comment by shimizukawa** [comment]
_From [timmie](https://bitbucket.org/timmie) on 2009-05-11 11:22:43+00:00_

Yes, it worked before.
I didn't change anythin on the Sphinx installation.

**Comment by shimizukawa** [comment]
_From [timmie](https://bitbucket.org/timmie) on 2009-05-11 12:17:39+00:00_

I just tred with:

import urllib2
handlers = [urllib2.ProxyHandler(), urllib2.HTTPRedirectHandler(),
            urllib2.HTTPHandler()]
if hasattr(urllib2, 'HTTPSHandler'):
    handlers.append(urllib2.HTTPSHandler)
urllib2.install_opener(urllib2.build_opener(*handlers))
f = urllib2.urlopen('http://sphinx.pocoo.org/latest/objects.inv')

=> 

<urlopen error (10060, 'Operation timed out')>

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-06-16 20:05:39+00:00_

And that may not have been caused by your proxy?

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-08-25 12:19:47+00:00_

Removing milestone: 0.6 (automated comment)

**Comment by donatelloOo** [comment]
loading intersphinx inventory from https://docs.python.org/objects.inv...
WARNING: intersphinx inventory 'https://docs.python.org/objects.inv' not fetchable due to <class 'requests.exceptions.SSLError'>: [SSL: TLSV1_ALERT_ACCESS_DENIED] tlsv1 alert access denied (_ssl.c:661)

InternetExplorer settings is configured to use enterprise proxy.

**Comment by tk0miya** [comment]
Since 1.5, Sphinx uses `requests` library for HTTP library.
So please use `HTTP_PROXY` and `HTTPS_PROXY` environment variables to set up your proxies.
http://docs.python-requests.org/en/master/user/advanced/#proxies

Thanks,

**Comment by espoelstra** [comment]
It appears that although Sphinx uses `requests` you may still need to set the certificate directory in your `conf.py` manually via a `tls_cacerts = '/path/to/your/certs'` as it doesn't appear to honor the `REQUESTS_CA_BUNDLE` or the other SSL environment variables commonly used to pass trusted certs. This has made using `sphinx-build linkcheck` very painful in an enterprise environment with a proxy because some of the resources also require a login to view, and hardcoding credentials anywhere is a bad idea.

**Comment by tk0miya** [comment]
@espoelstra Could you file a request as an issue please? I'll take a look later.

**Comment by ssbarnea** [comment]
This is still an issue and I can confirm that sphinx fails to load any of the known ca-bundle variables: ``SSL_CERT_FILE``, ``REQUESTS_CA_BUNDLE`` nor ``CURL_CA_BUNDLE``.

Anyone using a SSL proxy is pretty much ***.

**Comment by PhilippSelenium** [comment]
Why is this closed? Using ```tls_cacerts = <path_to_my_cert>``` still throws:
```
encountered some issues with some of the inventories, but they had working alternatives:
intersphinx inventory 'https://grpc.github.io/grpc/python/objects.inv' not fetchable due to <class 'requests.exceptions.SSLError'>: HTTPSConnectionPool(host='grpc.github.io', port=443): Max retries exceeded with url: /grpc/python/objects.inv (Caused by SSLError(SSLError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:852)'),))
```

**Comment by timhoffm** [comment]
Works for me either with `conf.py`

> tls_cacerts = '/path/to/your/certs.pem'

or with the environment variable

> export REQUESTS_CA_BUNDLE=/path/to/your/certs.pem

**Comment by beegeesquare** [comment]
@timhoffm I tried your method, but still the problem persists. I'm on corporate network proxy.  

`Warning, treated as error:
failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://docs.python.org/3/objects.inv' not fetchable due to <class 'requests.exceptions.SSLError'>: HTTPSConnectionPool(host='docs.python.org', port=443): Max retries exceeded with url: /3/objects.inv (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)')))`

Any ideas? I have created a certs.pem using `openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365` and exported it to the `bash_profile`. Is this the right way to create a `.pem` file? I'm on MacOS.

Can someone provide further tips on how to resolve this? I had been stuck with this error for months....unable to find a way to fix it ! Any help is highly appreciated.

Thanks !
